File: src/front-end/token.zig
//! token data type for lexical analysis
const std = @import("std");

/// a Token data type compromised of its type `token_type`,
/// its lexeme `value` for identifiers and literals, and
/// its position `pos` for compiler errors
pub const Token = struct {
    token_type: TokenType,
    value: ?[]const u8,

    /// create a new token object
    pub fn init(token_type: TokenType, value: ?[]const u8) Token {
        return Token{
            .token_type = token_type,
            .value = value,
        };
    }

    /// a token's type
    pub const TokenType = enum {
        /// identifier
        Identifier,

        /// literals
        IntegerLiteral,
        FloatLiteral,
        CharLiteral,
        StringLiteral,
        MultilineStringLiteral,

        /// delimiters
        LeftCircleBrack,
        RightCircleBrack,
        LeftCurlyBrack,
        RightCurlyBrack,
        LeftSquareBrack,
        RightSquareBrack,
        Colon,
        Semicolon,
        Dot,
        Comma,

        /// operators
        Equal,
        Bang,
        Plus,
        Minus,
        Asterisk,
        Ambersand,
        Slash,
        Percent,
        AmbersandEqual,
        Tilde,
        Pipe,
        PipeEqual,
        PlusEqual,
        MinusEqual,
        CaretEqual,
        Caret,
        AsteriskEqual,
        SlashEqual,
        PercentEqual,
        EqualEqual,
        BangEqual,
        RightAngleBracket,
        RightAngleBracketRightAngleBracketEqual,
        RightAngleBracketRightAngleBracket,
        RightAngleBracketEqual,
        LeftAngleBracket,
        LeftAngleBracketLeftAngleBracket,
        LeftAngleBracketLeftAngleBracketEqual,
        LeftAngleBracketEqual,

        SkinnyArrow,
        FatArrow,
        Ellipsis,
        EllipsisEqual,

        /// keywords
        Array,
        And,
        Bool,
        Break,
        Char,
        Const,
        Continue,
        Else,
        Enum,
        False,
        Field,
        Float,
        Fn,
        For,
        If,
        Implements,
        Import,
        In,
        Int,
        Interface,
        Match,
        Nil,
        Not,
        Or,
        Package,
        Pub,
        Return,
        String,
        Struct,
        This,
        True,
        Tuple,
        Var,
        Void,
        While,
        Step,

        /// special tokens
        Eof,
        Illegal,
    };

    /// a map to lookup if an identifier is actually a keyword
    pub const keyword_map = std.StaticStringMap(TokenType).initComptime(.{
        .{ "array", TokenType.Array },
        .{ "and", TokenType.And },
        .{ "bool", TokenType.Bool },
        .{ "break", TokenType.Break },
        .{ "char", TokenType.Char },
        .{ "const", TokenType.Const },
        .{ "continue", TokenType.Continue },
        .{ "else", TokenType.Else },
        .{ "enum", TokenType.Enum },
        .{ "false", TokenType.False },
        .{ "field", TokenType.Field },
        .{ "float", TokenType.Float },
        .{ "fn", TokenType.Fn },
        .{ "for", TokenType.For },
        .{ "if", TokenType.If },
        .{ "implements", TokenType.Implements },
        .{ "import", TokenType.Import },
        .{ "in", TokenType.In },
        .{ "int", TokenType.Int },
        .{ "interface", TokenType.Interface },
        .{ "match", TokenType.Match },
        .{ "nil", TokenType.Nil },
        .{ "not", TokenType.Not },
        .{ "or", TokenType.Or },
        .{ "package", TokenType.Package },
        .{ "pub", TokenType.Pub },
        .{ "return", TokenType.Return },
        .{ "string", TokenType.String },
        .{ "struct", TokenType.Struct },
        .{ "this", TokenType.This },
        .{ "true", TokenType.True },
        .{ "tuple", TokenType.Tuple },
        .{ "var", TokenType.Var },
        .{ "void", TokenType.Void },
        .{ "while", TokenType.While },
        .{ "step", TokenType.Step },
    });
};

File: src/front-end/token.c
#include "token.h"
#include <string.h>
#include <stdint.h>

typedef struct {
    const char *keyword;
    TokenType token_type;
} Entry;

uint32_t hash(const char *str) 
{
    uint32_t hash = 0;
    while (*str) {
        hash = hash * 31 + *str;
        str += 1;
    }
    return hash % 37;  // Use modulo of 37, the number of keywords
}

static Entry keyword_map[37] = {
    [3] = { keyword: "array", token_type: ARRAY },   
    [5] = { keyword: "and", token_type: AND },       
    [9] = { keyword: "bool", token_type: BOOL },     
    [12] = { keyword: "break", token_type: BREAK },  
    [16] = { keyword: "char", token_type: CHAR },    
    [18] = { keyword: "const", token_type: CONST },  
    [19] = { keyword: "continue", token_type: CONTINUE },
    [7] = { keyword: "else", token_type: ELSE },
    [10] = { keyword: "enum", token_type: ENUM },
    [11] = { keyword: "false", token_type: FALSE },
    [1] = { keyword: "field", token_type: FIELD },
    [8] = { keyword: "float", token_type: FLOAT },
    [20] = { keyword: "func", token_type: FUNC },
    [6] = { keyword: "for", token_type: FOR },
    [0] = { keyword: "if", token_type: IF },
    [13] = { keyword: "implements", token_type: IMPLEMENTS },
    [2] = { keyword: "import", token_type: IMPORT },
    [14] = { keyword: "in", token_type: IN },
    [22] = { keyword: "int", token_type: INT },
    [29] = { keyword: "interface", token_type: INTERFACE },
    [30] = { keyword: "match", token_type: MATCH },
    [4] = { keyword: "nil", token_type: NIL },
    [15] = { keyword: "not", token_type: NOT },
    [24] = { keyword: "or", token_type: OR },
    [17] = { keyword: "package", token_type: PACKAGE },
    [25] = { keyword: "pub", token_type: PUB },
    [23] = { keyword: "return", token_type: RETURN },
    [26] = { keyword: "string", token_type: STRING },
    [21] = { keyword: "struct", token_type: STRUCT },
    [27] = { keyword: "this", token_type: THIS },
    [28] = { keyword: "true", token_type: TRUE },
    [31] = { keyword: "tuple", token_type: TUPLE },
    [33] = { keyword: "var", token_type: VAR },
    [35] = { keyword: "void", token_type: VOID },
    [32] = { keyword: "while", token_type: WHILE },
    [34] = { keyword: "step", token_type: STEP }
};


TokenType lookup_keyword(const char *keyword) 
{
    size_t index = hash(keyword);
    if (keyword_map[index].keyword && strcmp(keyword_map[index].keyword, keyword) == 0) {
        return keyword_map[index].token;
    }
    return NULL;
}


